lib/spack/spack/binary_distribution.py

```diff
@@ -825,49 +841,55 @@ def get_spec(spec=None, force=False):
     return try_download_specs(urls=urls, force=force)
 
 
-def get_specs(force=False, allarch=False):
+def get_specs(allarch=False):
     """
     Get spec.yaml's for build caches available on mirror
     """
+    global _cached_specs
     arch = architecture.Arch(architecture.platform(),
                              'default_os', 'default_target')
-    arch_pattern = ('([^-]*-[^-]*-[^-]*)')
-    if not allarch:
-        arch_pattern = '(%s-%s-[^-]*)' % (arch.platform, arch.os)
-
-    regex_pattern = '%s(.*)(spec.yaml$)' % (arch_pattern)
-    arch_re = re.compile(regex_pattern)
 
     if not spack.mirror.MirrorCollection():
         tty.debug("No Spack mirrors are currently configured")
         return {}
 
-    urls = set()
     for mirror in spack.mirror.MirrorCollection().values():
         fetch_url_build_cache = url_util.join(
             mirror.fetch_url, _build_cache_relative_path)
 
-        mirror_dir = url_util.local_file_path(fetch_url_build_cache)
-        if mirror_dir:
-            tty.msg("Finding buildcaches in %s" % mirror_dir)
-            if os.path.exists(mirror_dir):
-                files = os.listdir(mirror_dir)
-                for file in files:
-                    m = arch_re.search(file)
-                    if m:
-                        link = url_util.join(fetch_url_build_cache, file)
-                        urls.add(link)
-        else:
-            tty.msg("Finding buildcaches at %s" %
-                    url_util.format(fetch_url_build_cache))
-            p, links = web_util.spider(
-                url_util.join(fetch_url_build_cache, 'index.html'))
-            for link in links:
-                m = arch_re.search(link)
-                if m:
-                    urls.add(link)
+        tty.msg("Finding buildcaches at %s" %
+                url_util.format(fetch_url_build_cache))
 
-    return try_download_specs(urls=urls, force=force)
+        index_url = url_util.join(fetch_url_build_cache, 'index.json')
+
+        try:
+            _, _, file_stream = web_util.read_from_url(
+                index_url, 'application/json')
+            index_object = codecs.getreader('utf-8')(file_stream).read()
+        except (URLError, web_util.SpackWebError) as url_err:
+            tty.error('Failed to read index {0}'.format(index_url))
+            tty.debug(url_err)
+            # Just return whatever specs we may already have cached
+            return _cached_specs
+
+        tmpdir = tempfile.mkdtemp()
+        index_file_path = os.path.join(tmpdir, 'index.json')
+        with open(index_file_path, 'w') as fd:
+            fd.write(index_object)
+
+        db_root_dir = os.path.join(tmpdir, 'db_root')
+        db = spack_db.Database(None, db_dir=db_root_dir,
+                               enable_transaction_locking=False)
```

Comment:
Should this pass the list of fields to expect too?

Reply:
I left it out because that's only used when writing, and the use case here is just to read in the index to a database instance so we can ask for the specs.

Reply:
> It looks like it's constructing a database that expects all of the normal fields, but without locks. That's different than what we use to write the file.

Hmm, yeah, I think I see what you're saying.  Maybe I should be providing that list here too, then using it to moderate the database class's expectations about what's present/required in the file it reads?

Reply:
That's what I was thinking. I think that's really the same issue through all three of my comments here. I'll mark the other two as resolved and we can discuss it here.

Reply:
What about instead of changing the `InstallRecord` type depending on the needed fields, change what type is used by the database. The default could be `InstallRecord` and the Database for a mirror could use a `BuildcacheRecord` type object (that just has fewer fields). And maybe both could inherit from `DatabaseRecord` to define the interface?

Then the constructor takes the class that it should use for records, and the `to_dict` and `from_dict` methods of the record could handle the changes.

But that might be too much of a refactor given that it would probably affect the Database.query API since we couldn't rely on certain fields being there.

Maybe we should merge this as-is (once you figure out the issue you're discussing with Omar below) and create a new issue to refactor this for a patch release? Since the refactor wouldn't be user-API facing we can give ourselves more time for it.

Reply:
Yeah, assuming the rest looks good to you, I vote for that.  Once I figure out what's going on with that test, you can merge this, and I'll write up the issue, link here, and ping you.  Thanks!